### ISSUE TITLE
jsonScheme to nickel converter

### ISSUE THREAD

**Issue by blaggacao** [tweag/nickel#418]
**Is your feature request related to a problem? Please describe.**
A lot of json schemata are readily available out there.

**Describe the solution you'd like**
A json scheme to nickel type system translator can help bootstrapping a lot of config schemata in nickel

**Describe alternatives you've considered**
Implement some sort of builtin `jsonSchemaValidator`

**Additional context**
none

**Comment by mtoohey31** [comment]
I gave something like this a try in the form of [a standalone program](https://github.com/mtoohey31/json-schema-to-nickel) that generates contract code given a JSON schema, but I've come to the conclusion that being able to fully translate JSON schemas into ergonomic Nickel contracts is impossible with the language in its current form because of the lack of support for union and intersection contracts (see [this related blog post](https://www.tweag.io/blog/2022-04-28-union-intersection-contracts/)). Since JSON schemas support the [allOf](https://json-schema.org/draft/2020-12/json-schema-core.html#name-allof) and [anyOf](https://json-schema.org/draft/2020-12/json-schema-core.html#name-anyof) keywords, but we don't support union and intersection, the best approach I could come up with was to turn the whole subschema into a predicate as soon as we encounter, for example, an `anyOf`. That leads to an absolutely unreadable contract and terrible blame messages though, so I didn't think that was worth implementing.

**Comment by yannham** [comment]
@mtoohey31 interesting experiment! 

I think intersection contracts alone are mostly fine to implement, as you long as you don't fiddle with functions. You can just apply them in sequence, being aware that this might not give what you think on schemas with functions inside (but I guess JSON schemas can't have that anyway). The issue comes from function contracts, which "negate" the argument, so you end up needing something like a union when you have an intersection for the domain contract.

Union is harder to handle. One brutal possibility would be to provide an operation such as in #714, which effectively allows you to turn an arbitrary contract into a predicate. That breaks the laziness unfortunately, but could be used as an escape hatch, because in the end it's probably still better to have JSON schemas although they are not lazy (as long as this is made very clear by the corresponding tool) rather than no interop at all.

My only fear is that people will abuse this operator to write all sort of unions, because this is just too handy, making in practice most contracts coming from a library potentially not lazy. Maybe a good compromise could be to at least wait to have a restricted `or` operation, as proposed in the blog post, which would hopefully fulfill most use-cases, before releasing such a `try/catch` like mechanism (which would be an ugly, unsafe primop like `%unsafe_catch%`)

**Comment by mtoohey31** [comment]
Thanks for the reply @yannham!

I went and read the section of the paper on intersection; I didn't realize the difficulties there were limited to functions, but that makes sense now. Based on that, something [like this](https://github.com/mtoohey31/json-schema-to-nickel/blob/84aa55973ac705a580b367094e5442727b8a8cc2/test3.ncl#L3) seems like it should behave correctly for the JSON case.

On the topic of #714, I looked at the JSON schema specification again from the perspective of figuring out what can be reasonably represented with contracts as they are today, and the [not](http://json-schema.org/draft/2020-12/json-schema-core.html#name-not) keyword also seems problematic. #714 would allow a `Not` contract to be defined quite easily, but I agree that the "escape hatch" approach could be risky because it might end up being overused. (If you do decide to implement something like that though, it seems like it might be useful to be able to attribute blame to the point where the `contract.validate` call failed, so maybe it could return something other than just a `Bool`?)

If we were to go the route of a restricted `or` operation instead, we would probably also need a similar `not` operator if we want to be able to fully support JSON schemas, so that we can cover all the basic boolean operations (and then we could get stuff like [oneOf](http://json-schema.org/draft/2020-12/json-schema-core.html#name-oneof), [if](http://json-schema.org/draft/2020-12/json-schema-core.html#name-if), etc. from combinations of the sequential, manually defined `All`, along with the restricted builtins `or` and `not`).

Either way, deciding what to do about unions is definitely a tricky issue...

**Comment by majidaldo** [comment]
...other way too:  nickel to json-schema.

**Comment by vkleen** [comment]
Since there continues to be interest in converting JSON schema and related specification into Nickel contracts, I've taken the liberty of updating @mtoohey31's first sketch to Nickel 1.0. I've also started experimenting with a library of predicate combinators as a first step towards converting all JSON schema constructs into predicates for Nickel.

I've also taken the liberty of copying @mtoohey31's code into the nickel-lang organisation at https://github.com/nickel-lang/json-schema-to-nickel.

@mtoohey31: I've invited you as a collaborator there and I hope you don't have objections to this.
